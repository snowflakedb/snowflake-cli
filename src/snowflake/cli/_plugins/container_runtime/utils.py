import enum
from dataclasses import dataclass
from pathlib import PurePath
from typing import Any, Dict, List, Optional, Union, cast

from packaging import version
from snowflake.snowpark import session


class SnowflakeCloudType(enum.Enum):
    AWS = "aws"
    AZURE = "azure"
    GCP = "gcp"

    @classmethod
    def from_value(cls, value: str) -> "SnowflakeCloudType":
        assert value
        for k in cls:
            if k.value == value.lower():
                return k
        else:
            raise ValueError(f"'{cls.__name__}' enum not found for '{value}'")


@dataclass(frozen=True)
class PayloadEntrypoint:
    file_path: PurePath
    main_func: Optional[str]


@dataclass(frozen=True)
class UploadedPayload:
    # TODO: Include manifest of payload files for validation
    stage_path: PurePath
    entrypoint: List[Union[str, PurePath]]


@dataclass(frozen=True)
class ComputeResources:
    cpu: float  # Number of vCPU cores
    memory: float  # Memory in GiB
    gpu: int = 0  # Number of GPUs
    gpu_type: Optional[str] = None


@dataclass(frozen=True)
class ImageSpec:
    repo: str
    image_name: str
    image_tag: str
    resource_requests: ComputeResources
    resource_limits: ComputeResources

    @property
    def full_name(self) -> str:
        return f"{self.repo}/{self.image_name}:{self.image_tag}"


def get_current_region_id(sess: session.Session) -> str:
    res = session.sql(sess, "SELECT CURRENT_REGION() AS CURRENT_REGION").collect()[0]

    return cast(str, res.CURRENT_REGION)


def get_current_snowflake_version(
    sess: session.Session, *, statement_params: Optional[Dict[str, Any]] = None
) -> version.Version:
    """Get Snowflake Version as a version.Version object follow PEP way of versioning, that is to say:
        "7.44.2 b202312132139364eb71238" to <Version('7.44.2+b202312132139364eb71238')>

    Args:
        sess: Snowpark Session.
        statement_params: Statement params. Defaults to None.

    Returns:
        The version of Snowflake Version.
    """
    res = session.sql(
        sess,
        "SELECT CURRENT_VERSION() AS CURRENT_VERSION",
        statement_params=statement_params,
    ).collect()[0]

    version_str = res.CURRENT_VERSION
    assert isinstance(version_str, str)

    version_str = "+".join(version_str.split())
    return version.parse(version_str)
