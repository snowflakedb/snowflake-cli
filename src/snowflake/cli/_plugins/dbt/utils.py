# Copyright (c) 2025 Snowflake Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from snowflake.cli._plugins.dbt.constants import KNOWN_SUBCOMMANDS


def _extract_dbt_args(args: list[str]) -> list[str]:
    flags = set()

    for arg in args:
        if arg.startswith("-"):
            if "=" in arg:
                flag_name = arg.split("=", 1)[0]
                flags.add(flag_name)
            else:
                flags.add(arg)
        elif arg in KNOWN_SUBCOMMANDS:
            flags.add(arg)

    return sorted(list(flags))
