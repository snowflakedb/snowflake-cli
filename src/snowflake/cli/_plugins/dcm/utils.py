# Copyright (c) 2024 Snowflake Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.


def format_test_failures(
    failed_expectations: list, total_tests: int, failed_count: int
) -> str:
    """Format test failures into a nice error message."""
    lines = [
        "Failed expectations:",
    ]

    for failed in failed_expectations:
        table_name = failed.get("table_name", "Unknown")
        expectation_name = failed.get("expectation_name", "Unknown")
        metric_name = failed.get("metric_name", "Unknown")
        expectation_expr = failed.get("expectation_expression", "N/A")
        value = failed.get("value", "N/A")

        lines.append(f"  Table: {table_name}")
        lines.append(f"  Expectation: {expectation_name}")
        lines.append(f"  Metric: {metric_name}")
        lines.append(f"  Expression: {expectation_expr}")
        lines.append(f"  Actual value: {value}")
        lines.append("")

    passed_tests = total_tests - failed_count
    lines.append(
        f"Tests completed: {passed_tests} passed, {failed_count} failed out of {total_tests} total."
    )

    return "\n".join(lines)


def format_refresh_results(refreshed_tables: list) -> str:
    """Format refresh results into a concise user-friendly message."""
    if not refreshed_tables:
        return "No dynamic tables found in the project."

    total_tables = len(refreshed_tables)
    refreshed_count = sum(
        1 for table in refreshed_tables if table.get("refreshed_dt_count", 0) > 0
    )
    up_to_date_count = total_tables - refreshed_count

    return f"{refreshed_count} dynamic table(s) refreshed. {up_to_date_count} dynamic table(s) up-to-date."
