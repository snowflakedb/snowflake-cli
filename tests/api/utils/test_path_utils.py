# Copyright (c) 2024 Snowflake Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
from __future__ import annotations

import os
from pathlib import Path

import pytest
from snowflake.cli.api.artifacts.common import NotInDeployRootError
from snowflake.cli.api.artifacts.utils import symlink_or_copy

from tests.nativeapp.utils import assert_dir_snapshot, touch
from tests.testing_utils.files_and_dirs import pushd, temp_local_dir
from tests_common import IS_WINDOWS


@pytest.mark.skipif(
    IS_WINDOWS, reason="Symlinks on Windows are restricted to Developer mode or admins"
)
def test_symlink_or_copy_raises_error(temp_dir, os_agnostic_snapshot):
    touch("GrandA/ParentA/ChildA")
    with open(Path(temp_dir, "GrandA/ParentA/ChildA"), "w") as f:
        f.write("Test 1")

    # Create the deploy root
    deploy_root = Path(temp_dir, "output", "deploy")
    os.makedirs(deploy_root)

    # Incorrect dst path
    with pytest.raises(NotInDeployRootError):
        symlink_or_copy(
            src=Path("GrandA", "ParentA", "ChildA"),
            dst=Path("output", "ParentA", "ChildA"),
            deploy_root=deploy_root,
        )

    file_in_deploy_root = Path("output", "deploy", "ParentA", "ChildA")

    # Correct path and parent directories are automatically created
    symlink_or_copy(
        src=Path("GrandA", "ParentA", "ChildA"),
        dst=file_in_deploy_root,
        deploy_root=deploy_root,
    )

    assert file_in_deploy_root.exists() and file_in_deploy_root.is_symlink()
    assert file_in_deploy_root.read_text(encoding="utf-8") == os_agnostic_snapshot

    # Since file_in_deploy_root is a symlink
    # it resolves to project_dir/GrandA/ParentA/ChildA, which is not in deploy root
    with pytest.raises(NotInDeployRootError):
        symlink_or_copy(
            src=Path("GrandA", "ParentA", "ChildA"),
            dst=file_in_deploy_root,
            deploy_root=deploy_root,
        )

    # Unlink the symlink file and create a file with the same name and path
    # This should pass since src.is_file() always begins by deleting the dst.
    os.unlink(file_in_deploy_root)
    touch(file_in_deploy_root)
    symlink_or_copy(
        src=Path("GrandA", "ParentA", "ChildA"),
        dst=file_in_deploy_root,
        deploy_root=deploy_root,
    )

    # dst is an existing symlink, will resolve to the src during NotInDeployRootError check.
    touch("GrandA/ParentA/ChildB")
    with pytest.raises(NotInDeployRootError):
        symlink_or_copy(
            src=Path("GrandA/ParentA/ChildB"),
            dst=file_in_deploy_root,
            deploy_root=deploy_root,
        )
    assert file_in_deploy_root.exists() and file_in_deploy_root.is_symlink()
    assert file_in_deploy_root.read_text(encoding="utf-8") == os_agnostic_snapshot


@pytest.mark.skipif(
    IS_WINDOWS, reason="Symlinks on Windows are restricted to Developer mode or admins"
)
def test_symlink_or_copy_with_no_symlinks_in_project_root(os_agnostic_snapshot):
    test_dir_structure = {
        "GrandA/ParentA/ChildA/GrandChildA": "Text GrandA/ParentA/ChildA/GrandChildA",
        "GrandA/ParentA/ChildA/GrandChildB.py": "Text GrandA/ParentA/ChildA/GrandChildB.py",
        "GrandA/ParentA/ChildA/GrandChildC": None,  # dir
        "GrandA/ParentA/ChildB.py": "Text GrandA/ParentA/ChildB.py",
        "GrandA/ParentA/ChildC": "Text GrandA/ParentA/ChildC",
        "GrandA/ParentA/ChildD": None,  # dir
        "GrandA/ParentB/ChildA": "Text GrandA/ParentB/ChildA",
        "GrandA/ParentB/ChildB.py": "Text GrandA/ParentB/ChildB.py",
        "GrandA/ParentB/ChildC/GrandChildA": None,  # dir
        "GrandA/ParentC": None,  # dir
        "GrandB/ParentA/ChildA": "Text GrandB/ParentA/ChildA",
        "output/deploy": None,  # dir
    }
    with temp_local_dir(test_dir_structure) as project_root:
        with pushd(project_root):
            # Sanity Check
            assert_dir_snapshot(Path("."), os_agnostic_snapshot)

            deploy_root = Path(project_root, "output/deploy")

            # "GrandB" dir
            symlink_or_copy(
                src=Path("GrandB/ParentA/ChildA"),
                dst=Path(deploy_root, "Grand1/Parent1/Child1"),
                deploy_root=deploy_root,
            )
            assert not Path(deploy_root, "Grand1").is_symlink()
            assert not Path(deploy_root, "Grand1/Parent1").is_symlink()
            assert Path(deploy_root, "Grand1/Parent1/Child1").is_symlink()

            # "GrandA/ParentC" dir
            symlink_or_copy(
                src=Path("GrandA/ParentC"),
                dst=Path(deploy_root, "Grand2"),
                deploy_root=deploy_root,
            )
            assert not Path(deploy_root, "Grand2").is_symlink()

            # "GrandA/ParentB" dir
            symlink_or_copy(
                src=Path("GrandA/ParentB/ChildA"),
                dst=Path(deploy_root, "Grand3"),
                deploy_root=deploy_root,
            )
            assert Path(deploy_root, "Grand3").is_symlink()
            symlink_or_copy(
                src=Path("GrandA/ParentB/ChildB.py"),
                dst=Path(deploy_root, "Grand4/Parent1.py"),
                deploy_root=deploy_root,
            )
            assert not Path(deploy_root, "Grand4").is_symlink()
            assert Path(deploy_root, "Grand4/Parent1.py").is_symlink()
            symlink_or_copy(
                src=Path("GrandA/ParentB/ChildC"),
                dst=Path(deploy_root, "Grand4/Parent2"),
                deploy_root=deploy_root,
            )
            assert not Path(deploy_root, "Grand4").is_symlink()
            assert not Path(deploy_root, "Grand4/Parent2").is_symlink()
            assert not Path(deploy_root, "Grand4/Parent2/GrandChildA").is_symlink()

            # "GrandA/ParentA" dir (1)
            symlink_or_copy(
                src=Path("GrandA/ParentA"), dst=deploy_root, deploy_root=deploy_root
            )
            assert not deploy_root.is_symlink()
            assert not Path(deploy_root, "ChildA").is_symlink()
            assert Path(deploy_root, "ChildA/GrandChildA").is_symlink()
            assert Path(deploy_root, "ChildA/GrandChildB.py").is_symlink()
            assert not Path(deploy_root, "ChildA/GrandChildC").is_symlink()
            assert Path(deploy_root, "ChildB.py").is_symlink()
            assert Path(deploy_root, "ChildC").is_symlink()
            assert not Path(deploy_root, "ChildD").is_symlink()

            # "GrandA/ParentA" dir (2)
            symlink_or_copy(
                src=Path("GrandA/ParentA"),
                dst=Path(deploy_root, "Grand4/Parent3"),
                deploy_root=deploy_root,
            )
            # Other children of Grand4 will be verified by a full assert_dir_snapshot(project_root) below
            assert not Path(deploy_root, "Grand4/Parent3").is_symlink()
            assert not Path(deploy_root, "Grand4/Parent3/ChildA").is_symlink()
            assert Path(deploy_root, "Grand4/Parent3/ChildA/GrandChildA").is_symlink()
            assert Path(
                deploy_root, "Grand4/Parent3/ChildA/GrandChildB.py"
            ).is_symlink()
            assert not Path(
                deploy_root, "Grand4/Parent3/ChildA/GrandChildC"
            ).is_symlink()
            assert Path(deploy_root, "Grand4/Parent3/ChildB.py").is_symlink()
            assert Path(deploy_root, "Grand4/Parent3/ChildC").is_symlink()
            assert not Path(deploy_root, "Grand4/Parent3/ChildD").is_symlink()

            assert_dir_snapshot(Path("./output/deploy"), os_agnostic_snapshot)

            # This is because the dst can be symlinks, which resolves to project src and hence outside deploy root.
            with pytest.raises(NotInDeployRootError):
                symlink_or_copy(
                    src=Path("GrandA/ParentB"),
                    dst=Path(deploy_root, "Grand4/Parent3"),
                    deploy_root=deploy_root,
                )


@pytest.mark.skipif(
    IS_WINDOWS, reason="Symlinks on Windows are restricted to Developer mode or admins"
)
def test_symlink_or_copy_with_symlinks_in_project_root(os_agnostic_snapshot):
    test_dir_structure = {
        "GrandA/ParentA": "Do not use as src of a symlink",
        "GrandA/ParentB": "Use as src of a symlink: GrandA/ParentB",
        "GrandA/ParentC/ChildA/GrandChildA": "Do not use as src of a symlink",
        "GrandA/ParentC/ChildA/GrandChildB": "Use as src of a symlink: GrandA/ParentC/ChildA/GrandChildB",
        "GrandB/ParentA/ChildA/GrandChildA": "Do not use as src of a symlink",
        "GrandB/ParentA/ChildB/GrandChildA": None,
        "symlinks/Grand1/Parent3/Child1": None,
        "symlinks/Grand2": None,
        "output/deploy": None,  # dir
    }
    with temp_local_dir(test_dir_structure) as project_root:
        with pushd(project_root):
            # Sanity Check
            assert_dir_snapshot(Path("."), os_agnostic_snapshot)

            os.symlink(
                Path("GrandA/ParentB").resolve(),
                Path(project_root, "symlinks/Grand1/Parent2"),
            )
            os.symlink(
                Path("GrandA/ParentC/ChildA/GrandChildB").resolve(),
                Path(project_root, "symlinks/Grand1/Parent3/Child1/GrandChild2"),
            )
            os.symlink(
                Path("GrandB/ParentA").resolve(),
                Path(project_root, "symlinks/Grand2/Parent1"),
                target_is_directory=True,
            )
            assert Path("symlinks").is_dir() and not Path("symlinks").is_symlink()
            assert (
                Path("GrandA/ParentB").is_file()
                and not Path("GrandA/ParentB").is_symlink()
            )
            assert (
                Path("symlinks/Grand1/Parent2").is_symlink()
                and Path("symlinks/Grand1/Parent2").is_file()
            )
            assert (
                Path("symlinks/Grand1/Parent3/Child1/GrandChild2").is_symlink()
                and Path("symlinks/Grand1/Parent3/Child1/GrandChild2").is_file()
            )
            assert (
                Path("symlinks/Grand2/Parent1").is_symlink()
                and Path("symlinks/Grand2/Parent1").is_dir()
            )

            # Sanity Check
            assert_dir_snapshot(Path("./symlinks"), os_agnostic_snapshot)

            deploy_root = Path(project_root, "output/deploy")

            symlink_or_copy(
                src=Path("GrandA"),
                dst=Path(deploy_root, "TestA"),
                deploy_root=deploy_root,
            )
            assert not Path(deploy_root, "TestA").is_symlink()
            assert Path(deploy_root, "TestA/ParentA").is_symlink()
            assert Path(deploy_root, "TestA/ParentB").is_symlink()
            assert not Path(deploy_root, "TestA/ParentC").is_symlink()
            assert not Path(deploy_root, "TestA/ParentC/ChildA").is_symlink()
            assert Path(deploy_root, "TestA/ParentC/ChildA/GrandChildA").is_symlink()
            assert Path(deploy_root, "TestA/ParentC/ChildA/GrandChildB").is_symlink()

            symlink_or_copy(
                src=Path("GrandB"),
                dst=Path(deploy_root, "TestB"),
                deploy_root=deploy_root,
            )
            assert not Path(deploy_root, "TestB").is_symlink()
            assert not Path(deploy_root, "TestB/ParentA").is_symlink()
            assert not Path(deploy_root, "TestB/ParentA/ChildA").is_symlink()
            assert not Path(deploy_root, "TestB/ParentA/ChildB").is_symlink()
            assert not Path(
                deploy_root, "TestB/ParentA/ChildB/GrandChildA"
            ).is_symlink()
            assert Path(deploy_root, "TestB/ParentA/ChildA/GrandChildA").is_symlink()

            symlink_or_copy(
                src=Path("symlinks"),
                dst=Path(deploy_root, "symlinks"),
                deploy_root=deploy_root,
            )
            assert (
                Path(deploy_root, "symlinks/Grand1").is_dir()
                and not Path(deploy_root, "symlinks/Grand1").is_symlink()
            )
            assert (
                Path(deploy_root, "symlinks/Grand1/Parent2").is_file()
                and Path(deploy_root, "symlinks/Grand1/Parent2").is_symlink()
            )
            assert (
                Path(deploy_root, "symlinks/Grand1/Parent3").is_dir()
                and not Path(deploy_root, "symlinks/Grand1/Parent3").is_symlink()
            )
            assert (
                Path(deploy_root, "symlinks/Grand1/Parent3/Child1").is_dir()
                and not Path(deploy_root, "symlinks/Grand1/Parent3/Child1").is_symlink()
            )
            assert (
                Path(
                    deploy_root, "symlinks/Grand1/Parent3/Child1/GrandChild2"
                ).is_file()
                and Path(
                    deploy_root, "symlinks/Grand1/Parent3/Child1/GrandChild2"
                ).is_symlink()
            )
            assert (
                Path(deploy_root, "symlinks/Grand2").is_dir()
                and not Path(deploy_root, "symlinks/Grand2").is_symlink()
            )
            assert (
                Path(deploy_root, "symlinks/Grand2/Parent1").is_dir()
                and not Path(deploy_root, "symlinks/Grand2/Parent1").is_symlink()
            )
            assert (
                Path(deploy_root, "symlinks/Grand2/Parent1/ChildA").is_dir()
                and not Path(deploy_root, "symlinks/Grand2/Parent1/ChildA").is_symlink()
            )
            assert (
                Path(
                    deploy_root, "symlinks/Grand2/Parent1/ChildA/GrandChildA"
                ).is_file()
                and Path(
                    deploy_root, "symlinks/Grand2/Parent1/ChildA/GrandChildA"
                ).is_symlink()
            )
            assert (
                Path(deploy_root, "symlinks/Grand2/Parent1/ChildB/GrandChildA").is_dir()
                and not Path(
                    deploy_root, "symlinks/Grand2/Parent1/ChildB/GrandChildA"
                ).is_symlink()
            )

            assert_dir_snapshot(Path("./output/deploy"), os_agnostic_snapshot)
