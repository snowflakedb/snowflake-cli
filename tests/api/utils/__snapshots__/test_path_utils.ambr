# serializer version: 1
# name: test_symlink_or_copy_raises_error
  'Test 1'
# ---
# name: test_symlink_or_copy_raises_error.1
  'Test 1'
# ---
# name: test_symlink_or_copy_with_no_symlinks_in_project_root
  '''
  d .
  d GrandA
  d GrandA/ParentA
  d GrandA/ParentA/ChildA
  f GrandA/ParentA/ChildA/GrandChildA
  f GrandA/ParentA/ChildA/GrandChildB.py
  d GrandA/ParentA/ChildA/GrandChildC
  f GrandA/ParentA/ChildB.py
  f GrandA/ParentA/ChildC
  d GrandA/ParentA/ChildD
  d GrandA/ParentB
  f GrandA/ParentB/ChildA
  f GrandA/ParentB/ChildB.py
  d GrandA/ParentB/ChildC
  d GrandA/ParentB/ChildC/GrandChildA
  d GrandA/ParentC
  d GrandB
  d GrandB/ParentA
  f GrandB/ParentA/ChildA
  d output
  d output/deploy
  '''
# ---
# name: test_symlink_or_copy_with_no_symlinks_in_project_root.1
  '''
  ===== Contents of: GrandA/ParentA/ChildA/GrandChildA =====
  Text GrandA/ParentA/ChildA/GrandChildA
  '''
# ---
# name: test_symlink_or_copy_with_no_symlinks_in_project_root.10
  '''
  ===== Contents of: output/deploy/ChildA/GrandChildB.py =====
  Text GrandA/ParentA/ChildA/GrandChildB.py
  '''
# ---
# name: test_symlink_or_copy_with_no_symlinks_in_project_root.11
  '''
  ===== Contents of: output/deploy/ChildB.py =====
  Text GrandA/ParentA/ChildB.py
  '''
# ---
# name: test_symlink_or_copy_with_no_symlinks_in_project_root.12
  '''
  ===== Contents of: output/deploy/ChildC =====
  Text GrandA/ParentA/ChildC
  '''
# ---
# name: test_symlink_or_copy_with_no_symlinks_in_project_root.13
  '''
  ===== Contents of: output/deploy/Grand1/Parent1/Child1 =====
  Text GrandB/ParentA/ChildA
  '''
# ---
# name: test_symlink_or_copy_with_no_symlinks_in_project_root.14
  '''
  ===== Contents of: output/deploy/Grand3 =====
  Text GrandA/ParentB/ChildA
  '''
# ---
# name: test_symlink_or_copy_with_no_symlinks_in_project_root.15
  '''
  ===== Contents of: output/deploy/Grand4/Parent1.py =====
  Text GrandA/ParentB/ChildB.py
  '''
# ---
# name: test_symlink_or_copy_with_no_symlinks_in_project_root.16
  '''
  ===== Contents of: output/deploy/Grand4/Parent3/ChildA/GrandChildA =====
  Text GrandA/ParentA/ChildA/GrandChildA
  '''
# ---
# name: test_symlink_or_copy_with_no_symlinks_in_project_root.17
  '''
  ===== Contents of: output/deploy/Grand4/Parent3/ChildA/GrandChildB.py =====
  Text GrandA/ParentA/ChildA/GrandChildB.py
  '''
# ---
# name: test_symlink_or_copy_with_no_symlinks_in_project_root.18
  '''
  ===== Contents of: output/deploy/Grand4/Parent3/ChildB.py =====
  Text GrandA/ParentA/ChildB.py
  '''
# ---
# name: test_symlink_or_copy_with_no_symlinks_in_project_root.19
  '''
  ===== Contents of: output/deploy/Grand4/Parent3/ChildC =====
  Text GrandA/ParentA/ChildC
  '''
# ---
# name: test_symlink_or_copy_with_no_symlinks_in_project_root.2
  '''
  ===== Contents of: GrandA/ParentA/ChildA/GrandChildB.py =====
  Text GrandA/ParentA/ChildA/GrandChildB.py
  '''
# ---
# name: test_symlink_or_copy_with_no_symlinks_in_project_root.3
  '''
  ===== Contents of: GrandA/ParentA/ChildB.py =====
  Text GrandA/ParentA/ChildB.py
  '''
# ---
# name: test_symlink_or_copy_with_no_symlinks_in_project_root.4
  '''
  ===== Contents of: GrandA/ParentA/ChildC =====
  Text GrandA/ParentA/ChildC
  '''
# ---
# name: test_symlink_or_copy_with_no_symlinks_in_project_root.5
  '''
  ===== Contents of: GrandA/ParentB/ChildA =====
  Text GrandA/ParentB/ChildA
  '''
# ---
# name: test_symlink_or_copy_with_no_symlinks_in_project_root.6
  '''
  ===== Contents of: GrandA/ParentB/ChildB.py =====
  Text GrandA/ParentB/ChildB.py
  '''
# ---
# name: test_symlink_or_copy_with_no_symlinks_in_project_root.7
  '''
  ===== Contents of: GrandB/ParentA/ChildA =====
  Text GrandB/ParentA/ChildA
  '''
# ---
# name: test_symlink_or_copy_with_no_symlinks_in_project_root.8
  '''
  d output/deploy
  d output/deploy/ChildA
  f output/deploy/ChildA/GrandChildA
  f output/deploy/ChildA/GrandChildB.py
  d output/deploy/ChildA/GrandChildC
  f output/deploy/ChildB.py
  f output/deploy/ChildC
  d output/deploy/ChildD
  d output/deploy/Grand1
  d output/deploy/Grand1/Parent1
  f output/deploy/Grand1/Parent1/Child1
  d output/deploy/Grand2
  f output/deploy/Grand3
  d output/deploy/Grand4
  f output/deploy/Grand4/Parent1.py
  d output/deploy/Grand4/Parent2
  d output/deploy/Grand4/Parent2/GrandChildA
  d output/deploy/Grand4/Parent3
  d output/deploy/Grand4/Parent3/ChildA
  f output/deploy/Grand4/Parent3/ChildA/GrandChildA
  f output/deploy/Grand4/Parent3/ChildA/GrandChildB.py
  d output/deploy/Grand4/Parent3/ChildA/GrandChildC
  f output/deploy/Grand4/Parent3/ChildB.py
  f output/deploy/Grand4/Parent3/ChildC
  d output/deploy/Grand4/Parent3/ChildD
  '''
# ---
# name: test_symlink_or_copy_with_no_symlinks_in_project_root.9
  '''
  ===== Contents of: output/deploy/ChildA/GrandChildA =====
  Text GrandA/ParentA/ChildA/GrandChildA
  '''
# ---
# name: test_symlink_or_copy_with_symlinks_in_project_root
  '''
  d .
  d GrandA
  f GrandA/ParentA
  f GrandA/ParentB
  d GrandA/ParentC
  d GrandA/ParentC/ChildA
  f GrandA/ParentC/ChildA/GrandChildA
  f GrandA/ParentC/ChildA/GrandChildB
  d GrandB
  d GrandB/ParentA
  d GrandB/ParentA/ChildA
  f GrandB/ParentA/ChildA/GrandChildA
  d GrandB/ParentA/ChildB
  d GrandB/ParentA/ChildB/GrandChildA
  d output
  d output/deploy
  d symlinks
  d symlinks/Grand1
  d symlinks/Grand1/Parent3
  d symlinks/Grand1/Parent3/Child1
  d symlinks/Grand2
  '''
# ---
# name: test_symlink_or_copy_with_symlinks_in_project_root.1
  '''
  ===== Contents of: GrandA/ParentA =====
  Do not use as src of a symlink
  '''
# ---
# name: test_symlink_or_copy_with_symlinks_in_project_root.10
  '''
  ===== Contents of: output/deploy/TestA/ParentA =====
  Do not use as src of a symlink
  '''
# ---
# name: test_symlink_or_copy_with_symlinks_in_project_root.11
  '''
  ===== Contents of: output/deploy/TestA/ParentB =====
  Use as src of a symlink: GrandA/ParentB
  '''
# ---
# name: test_symlink_or_copy_with_symlinks_in_project_root.12
  '''
  ===== Contents of: output/deploy/TestA/ParentC/ChildA/GrandChildA =====
  Do not use as src of a symlink
  '''
# ---
# name: test_symlink_or_copy_with_symlinks_in_project_root.13
  '''
  ===== Contents of: output/deploy/TestA/ParentC/ChildA/GrandChildB =====
  Use as src of a symlink: GrandA/ParentC/ChildA/GrandChildB
  '''
# ---
# name: test_symlink_or_copy_with_symlinks_in_project_root.14
  '''
  ===== Contents of: output/deploy/TestB/ParentA/ChildA/GrandChildA =====
  Do not use as src of a symlink
  '''
# ---
# name: test_symlink_or_copy_with_symlinks_in_project_root.15
  '''
  ===== Contents of: output/deploy/symlinks/Grand1/Parent2 =====
  Use as src of a symlink: GrandA/ParentB
  '''
# ---
# name: test_symlink_or_copy_with_symlinks_in_project_root.16
  '''
  ===== Contents of: output/deploy/symlinks/Grand1/Parent3/Child1/GrandChild2 =====
  Use as src of a symlink: GrandA/ParentC/ChildA/GrandChildB
  '''
# ---
# name: test_symlink_or_copy_with_symlinks_in_project_root.17
  '''
  ===== Contents of: output/deploy/symlinks/Grand2/Parent1/ChildA/GrandChildA =====
  Do not use as src of a symlink
  '''
# ---
# name: test_symlink_or_copy_with_symlinks_in_project_root.2
  '''
  ===== Contents of: GrandA/ParentB =====
  Use as src of a symlink: GrandA/ParentB
  '''
# ---
# name: test_symlink_or_copy_with_symlinks_in_project_root.3
  '''
  ===== Contents of: GrandA/ParentC/ChildA/GrandChildA =====
  Do not use as src of a symlink
  '''
# ---
# name: test_symlink_or_copy_with_symlinks_in_project_root.4
  '''
  ===== Contents of: GrandA/ParentC/ChildA/GrandChildB =====
  Use as src of a symlink: GrandA/ParentC/ChildA/GrandChildB
  '''
# ---
# name: test_symlink_or_copy_with_symlinks_in_project_root.5
  '''
  ===== Contents of: GrandB/ParentA/ChildA/GrandChildA =====
  Do not use as src of a symlink
  '''
# ---
# name: test_symlink_or_copy_with_symlinks_in_project_root.6
  '''
  d symlinks
  d symlinks/Grand1
  f symlinks/Grand1/Parent2
  d symlinks/Grand1/Parent3
  d symlinks/Grand1/Parent3/Child1
  f symlinks/Grand1/Parent3/Child1/GrandChild2
  d symlinks/Grand2
  d symlinks/Grand2/Parent1
  '''
# ---
# name: test_symlink_or_copy_with_symlinks_in_project_root.7
  '''
  ===== Contents of: symlinks/Grand1/Parent2 =====
  Use as src of a symlink: GrandA/ParentB
  '''
# ---
# name: test_symlink_or_copy_with_symlinks_in_project_root.8
  '''
  ===== Contents of: symlinks/Grand1/Parent3/Child1/GrandChild2 =====
  Use as src of a symlink: GrandA/ParentC/ChildA/GrandChildB
  '''
# ---
# name: test_symlink_or_copy_with_symlinks_in_project_root.9
  '''
  d output/deploy
  d output/deploy/TestA
  f output/deploy/TestA/ParentA
  f output/deploy/TestA/ParentB
  d output/deploy/TestA/ParentC
  d output/deploy/TestA/ParentC/ChildA
  f output/deploy/TestA/ParentC/ChildA/GrandChildA
  f output/deploy/TestA/ParentC/ChildA/GrandChildB
  d output/deploy/TestB
  d output/deploy/TestB/ParentA
  d output/deploy/TestB/ParentA/ChildA
  f output/deploy/TestB/ParentA/ChildA/GrandChildA
  d output/deploy/TestB/ParentA/ChildB
  d output/deploy/TestB/ParentA/ChildB/GrandChildA
  d output/deploy/symlinks
  d output/deploy/symlinks/Grand1
  f output/deploy/symlinks/Grand1/Parent2
  d output/deploy/symlinks/Grand1/Parent3
  d output/deploy/symlinks/Grand1/Parent3/Child1
  f output/deploy/symlinks/Grand1/Parent3/Child1/GrandChild2
  d output/deploy/symlinks/Grand2
  d output/deploy/symlinks/Grand2/Parent1
  d output/deploy/symlinks/Grand2/Parent1/ChildA
  f output/deploy/symlinks/Grand2/Parent1/ChildA/GrandChildA
  d output/deploy/symlinks/Grand2/Parent1/ChildB
  d output/deploy/symlinks/Grand2/Parent1/ChildB/GrandChildA
  '''
# ---
